Pipeline Update

LouisLu705 opened the pull request:
# Description

This pipeline change implements processing CSV files for meter and readings data in one place. This new process will check that the readings data does not violate expected constraints and outputs warnings as necessary if errors occur. Currently, this is tested on MAMAC data and works for CSV files which may come with several parameters to indicate how OED should process a certain CSV file such as: 

onlyEndTime - CSV files with onlyEndTime values given
cumulative - CSV files with readings which are cumulative
Tgap - the allowed time gap in milliseconds that a gap may occur between two readings 
Tlen - the allowed time variation in milliseconds that two readings may deviate from each other
cumulativeReset - true if the cumulative data can reset
resetStart - the start time of the allowed time range a cumulative CSV file may reset
resetEnd -  the end time of the allowed time range a cumulative CSV file may reset


Fixes #513 and Fixes #546

See @huss comments below on changes that fixes #510, fixes #512, fixes #515, fixes #565 and fixes #641.

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

This merges in the [csvPipeline](https://github.com/lindavin/OED/tree/csvPipeline). branch creataed by @lindavin. 

Currently, we expect that dates are passed in as one of two formats:

(YYYY-MM-DD H:mm) or as (MM-DD-YYYY H:mm) == (2020/4/5 0:30) or as (4/5/2020 0:30)

If the dates are in neither of these forms there may be errors. There is also a slight performance decrease due to checking both formats so if we can default to one specific moment format we may improve performance slightly. The difference I found was 11 seconds longer when testing on OneMinuteTestingData.

Also, we need to connect the user interface to the pipeline so that users can decide if their data has certain parameters such as specific cumulativeResetStart and cumulativeResetEnd times as well as onlyEndTime. Note that currently onlyEndTime is set to be false and if no cumulatveResetStart or cumulativeResetEnd times are given the values respectively default to 0:00:00.000 and 23:59:99.999 indicating that a cumulativeReset may occur from the beginning of any day to the millisecond before the day ends and the next day begins.

We may also consider counting how many errors occur and the client may use a parameter to create a maximum cap before too many errors occur and we drop the CSV file and do not insert anything into the database.

Also, we may make the error messages more descriptive for example letting the user know which reading causes the error including that particular reading value, meter id, start timestamp and end timestamp.

There is now also redundant code in the pipeline that will need to be cleared and testing will need to be done to see if we can use this new pipeline process to handle different types of meters besides Mamac such as Obvius. 

Here are some statistics from running the generatedTestData on my local machine with 8 GB of RAM.

fourDayFreqTestData:

    real    0m0.395s
    user    0m0.087s
    sys     0m0.095s

generateFourHourTestingData:

    real    0m0.922s
    user    0m0.007s
    sys     0m0.006s
generateTwentyThreeMinuteTestingData:

    real    0m5.593s
    user    0m0.000s
    sys     0m0.017s
generateFifteenMinuteTestingData:

    real    0m8.075s
    user    0m0.010s
    sys     0m0.011s
generateCosineTestingData:

    real    0m5.260s
    user    0m0.015s
    sys     0m0.000s
generateSineSquaredTestingData:

    real    0m0.181s
    user    0m0.000s
    sys     0m0.013s
generateCosineSquaredTestingData:

    real    0m0.149s
    user    0m0.000s
    sys     0m0.011s
generateVariableAmplitudeTestingData:

 - 15Freq1AmpSineTestData:
 
        real    0m16.320s
        user    0m0.000s
        sys     0m0.043s
 - 15Freq2AmpSineTestData:
 
        real    0m16.065s
        user    0m0.001s
        sys     0m0.048s
 - 15Freq3AmpSineTestData:
 
        real    0m16.329s
        user    0m0.010s
        sys     0m0.036s
 - 15Freq4AmpSineTestData:
 
        real    0m15.876s
        user    0m0.012s
        sys     0m0.034s
 - 15Freq5AmpSineTestData:
 
        real    0m15.612s
        user    0m0.001s
        sys     0m0.046s
 - 15Freq6AmpSineTestData:
   
        real    0m16.351s
        user    0m0.011s
        sys     0m0.031s
 - 15Freq7AmpSineTestData:
 
        real    0m15.882s
        user    0m0.010s
        sys     0m0.035s

generateOneMinuteTestingData:

	real    2m1.651s
	user    0m0.020s
	sys     0m0.222s

LouisLu705 committed:
Merged lindavins csvPipeline branch.

LouisLu705 committed:
Initial merge of Davin's csvPipeline and beginning implementation of processData.

LouisLu705 committed:
Added specifications and implementation of processData. Also removed unecessary debug statements.

LouisLu705 committed:
added test files

LouisLu705 committed:
Finished first implementation of processData.

LouisLu705 committed:
Added functionality to handleCumulativeReset

LouisLu705 committed:
Added default values for cumulativeResetStart and cumulativeResetEnd to be any time.

LouisLu705 committed:
fixed formatting so moment date is the format Y/M/D H:mm

LouisLu705 committed:
Added generatedTestFileData

LouisLu705 committed:
Matched new moment format where necessary.

LouisLu705 committed:
Added functionality to handle two moment formats: YYYY-MM-DD and MM-DD-YYYY

LouisLu705 committed:
Added some more specifications to paramters. Also removed unused imports.

LouisLu705 committed:
Added some more comments on what we expect readings to look like.

LouisLu705 committed:
Merged Davin's newest version of csvPipeline.

LouisLu705 committed:
removed generatedTestFiles

LouisLu705 committed:
Fixed format to agree with npm indents, indents, max string length, and single quotes.

LouisLu705 committed:
removed unecessary folder.

huss commented on src/server/services/pipeline-in-progress/handleCumulativeReset.js:
```diff
@@ -0,0 +1,39 @@
+const moment = require('moment');
+
+/**
+ * @param {boolean} cumulativeReset true if the client expects to handle cumulativeReset
```
It might be nice to tell the allowed date/time format in the JSDoc.

LouisLu705 replied:
For clarification @huss do you mean the allowed date/time format of the readings data or the allowed date/time format that we expect reset time specifications to be passed in as or maybe both?

huss replied:
I was thinking of the reset time since that is an argument to this function but more documentation is fine :-)

LouisLu705 replied:
Done.

huss commented on src/server/services/pipeline-in-progress/handleCumulativeReset.js:
```diff
@@ -0,0 +1,39 @@
+		return false;
+	}
+	else {
+		let momentFormat = '';
```
I think some comments here would help. The fact you are choosing between formats (I think) and what the moment creation is doing below with that format.

LouisLu705 replied:
Noted and done.

huss commented on src/server/services/pipeline-in-progress/loadArrayInput.js:
```diff
@@ -22,11 +19,18 @@ const handleCumulative = require('./handleCumulative');
-	readings = convertToReadings(readingsArray, meterID, conditionSet);
-	return await Reading.insertOrIgnoreAll(readings, conn);
+
+	// Temporary values for params
```
Are these placeholders until the pipeline is able to pass them along? If so, maybe a TODO comment and a function declaration with the expected parameters with these default values?

LouisLu705 replied:
Done.

huss commented on src/server/test/web/csvPipeline/README:
```diff
@@ -0,0 +1,4 @@
+This folder is need by src/server/data/automatedTestingData.js for placing
```
Should this be in the directory db/data/automatedTests? It is that directory that needs to be present (I think).

Also, there are a number of csv files in that directory that were added. I believe they came from running the automated test data script. We were not planning to provide with OED and store in git but have people generate locally as needed. Thus, I think they should be removed unless there is a new usage I am not aware of.

LouisLu705 replied:
Thank you for this catch. This entire automatedTests directory should not be in there. Instead, csv pipeline test files should be in the src/server/services/test/web/csvPipeline directory.

huss commented on src/server/services/csvPipeline/uploadReadings.js:
```diff
@@ -0,0 +1,54 @@
+		conn
+	); // load csv data
+	// TODO: If unsuccessful upload then an error will be thrown. We need to catch this error.
+	require ('../refreshReadingViews');
```
First, a comment here about the refreshReadingViews would be helpful. Second, I thought there was going to be a parameter to decide if a refresh should happen.

LouisLu705 replied:
Done.

Edit: I've noticed after some more testing that this functionality actually was recently added to a different part of the pipeline. So I am removing th refreshReadingViews from UploadReadings since it is redundant.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+ * @param {number} Tlen the allowed time variation in milliseconds that two readings may deviate from each other
+ * @param {dict} conditionSet used to validate readings (minVal, maxVal, minDate, maxDate, interval, maxError)
+ */
+
```
remove blank line.

LouisLu705 replied:
Done and noted.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+function processData(rows, meterID, isCumulative, cumulativeReset, resetStart = '0:00:00.000', 
+					resetEnd = '23:59:99.999', readingRepetition, onlyEndTime = false, Tgap, Tlen, conditionSet) {
+
+	// If processData is succesfully finished then return result = [R0, R1, R2...RN]
```
successfully

LouisLu705 replied:
Corrected the typo thank you.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+const Reading = require('../../models/Reading');
+const handleCumulativeReset = require('./handleCumulativeReset');
+
+
```
remove second blank line.

LouisLu705 replied:
Done and noted.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+	}
+	return result;
+}
+function isFirst(t) {
```
Blank line before function. It should also have JSDoc.

LouisLu705 replied:
I've added specifications and added a line thank you.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+ * @param {string} resetEnd a string representation of the end time a cumulativeReset may occur before. The default resetEnd time is 23:59:99
+ * @param readingRepetition value is 1 if reading is not duplicated. 2 if repeated twice and so on (E-mon D-mon meters)
+ * @param {boolean} onlyEndTime true if the data only has an endTimestamp
+ * @param {number} Tgap the allowed time variation in milliseconds that a gap may occur between two readings 
```
I'm thinking it makes more sense to make Tgap & Tlen be in seconds. The code can convert to milliseconds. The user can give values past the decimal if they want less than a second.

LouisLu705 replied:
Ok the logic accepts ms due to how moment calculates a moment.diff(). Therefore I've gone ahead and just multiplied the value passed in by 1000. This should work if not then we may need to dig deeper.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
+	let startTimestamp = moment(0);
```
Maybe comment on why this initial value.

LouisLu705 replied:
I've added comments explaining this.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+	let startTimestamp = moment(0);
+	let endTimestamp = moment(0);
+
+	let currentReading = new Reading(meterID, -1, startTimestamp, endTimestamp);
```
I think a comment on what current and previous readings represent and how used would be very helpful.

LouisLu705 replied:
I've added a large comment explaining how these values are used and calculated. It should be clear but if not I will revise as deemed necessary.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+	return result;
+}
+function isFirst(t) {
+	const E0 = moment(0);
```
I think it might be better to declare this constant outside both functions and then use it here and to initialize the first times above.

LouisLu705 replied:
I agree. I've made it a constant value outside of the function.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+	let meterReading = 0;
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
```
I think the plan was to use a value stored on the meter as the first previous reading. If this is not done because the DB is not ready then it would be good to have TODO items for each place that a change will be needed.

LouisLu705 replied:
@huss Could you elaborate with a small example? I'm confused by what "the value stored on the meter" means. Is this the last reading value for a particular meter id or something else?

huss replied:
The recently merged PR #632 added attributes to the meters table. They include reading, start_timestamp & end_timestamp. These three values are supposed to represent the last reading stored for that meter from the pipeline. As such, they should be the first value used as the previous reading for the next value(s) that come to the pipeline for that meter. The last value processed by the pipeline should always be stored back to the DB so it is ready for the next batch. Note the reading for cumulative data is the raw, cumulative value so it still works for the next time. (The value in the readings table will be the net usages for that time frame so it will differ.) The default value when a meter is created sets the value to 0, 0001-01-01 00:00:00, 0001-01-01 00:00:00. The plan was that such a date/time (so far back in time) would mean the pipeline would still work correctly for the first point. It will generate a warning due to the time gap and not use the first ever cumulative value due to that gap but that is what is desired. Note the Meter.js model should now return these values so it can be used to get these three values from the DB and store new ones back. We really should put them into Redux with other meter state but we can skip that for now if you want. Just leave a TODO if you do that. Let  me know if this is not clear or the default values do no work as hoped.

LouisLu705 replied:
I think I've completed this. I've left a TODO for the redux state since I'm not sure how to go about doing this.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing with cumulative data.';
+			}
+			if (onlyEndTime && isFirst(prevReading.endTimestamp)) {
```
Should this be an else if since the first point may have been dropped from cumulative? If not, then the errMsg should probably be +=.

LouisLu705 replied:
Done and noted. The way it's done now means that if the data is the first ever meter readings passed in for cumulative data and onlyEndTimes are used then we append the first drop error message so the user sees: 

"The first reading must be dropped when dealing with cumulative data. The first reading must be dropped when dealing with only endTimestamps."

I've also considered potentially just using the first one so that user doesn't see the second sentence but I think it makes more sense that the user sees as much information as possible. This can be changed if deemed necessary.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+			}
+			if (!readingOK) {
+				// An error occurred, for now log it, let the user know and continue
+				log.error(errMsg);
```
I think it would be good to give the reading values in the errMsg by adding here. Also, nice if told the reading number so easier to find in the file.

LouisLu705 replied:
I've gone and done this. The user will now see:

"Error parsing Reading # {}. Reading value gives {} with error message: {}".

The first curly brace is the row that the user will see the reading such as in the csv file. the second curly brace is the actual meter reading value and then the last curly brace is the error message on why this reading was dropped.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
```
I don't know if you need the second sentence since the first reading is excluded by the if. Should probably add that only warning since possible reading are not sent in time order.

LouisLu705 replied:
You're right. I've changed it so that the second sentence was removed and the user will instead see:

"The current reading startTime is not after the previous reading's end time. Note this is treated only as a warning since readings may be sent out of order"

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				}
+				// Reject negative readings
+				if (isCumulative && meterReading1 < 0) {
+					// If the value isCumulative and the first reading is negative reject unless cumulativeReset is true
```
I think it might be confusing to future people because "first reading" is used in two ways. Above, when it is the first reading ever received and here where it is the prior reading for cumulative data. I might call this one prior.

LouisLu705 replied:
I agree it was confusing and might be difficult for future developers. I've added more comment which hopefully make it clear what each meterReading value actually represents. 

So for example in the case of cumulative,

meterReading1 // In cumulative this is the previous raw reading
meterReading2 // In cumulative this is the current raw reading
meterReading // This is the net reading value.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+					result.push(currentReading);
+				}
+			}
+			if (!readingOK) {
```
I think this can be an else since it could be the follow up to if (readingOK) above.

LouisLu705 replied:
This is actually a little tricky because even though the final readingOK may be true it might be the case that the meterReading still errors out in which case the reading is dropped. 

So for example start times and end times might pass all checks and drop into if (readingOK). However, at this point we might find out that meterReading turns out to be negative and cumulative in which we case we must drop the reading and we still want to let the user know that the reading dropped even though it was ok up until this point. 

That's why I still think it's correct to leave it as an if !readingOK since if it occurs that readingOK is not true there is an accompanying error message that we should always let the client know,

HOWEVER, with the new changes to the meter model I've changed the logic so that meterReading is calculated along with start and end times in the beginning so it might make sense to move the checks out of the current if block and move it above the current if block. If we do this we can definitely do all the checks regarding meterReadings, start timestamps and end timestamps and then finally there are only two possibilities, either there is an error (readingOK == false )or no error (readingOK == true) in which case we can use an if else block. I hope this is clear but if it's not I can definitely try to make it clearer.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+			prevReading = currentReading;
+		}
+	}
+	if (conditionSet !== undefined && !validateReadings(result, conditionSet)) {
```
I'm unsure how this function is being used since I think it comes from ./validateReadings.js but is not in a require statement. Not sure what I am missing.

LouisLu705 replied:
Apologies this an oversight. I forgot to import this js file. I've updated it so it should work correctly now. It should be used to check that the final result array which contains all the readings is well-formed.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
```
Should probably include that this is only a warning. Also, maybe += since I cannot convince myself this is independent from the warning above.

LouisLu705 replied:
I've gone ahead and done this. However, I have some doubts because it currently never calls this warning if the reading is the first reading which is the only case (I think) we expect an unexpected gap.

Therefore, I think this warning is only called in the case there is an actual unexpected gap in readings which I think is a valid error that we might want to consider doing something with.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
```
It is possible the errMsg is not empty but readingOK is true. We need to log that in this case. May need to do here or use more complex logic since there is one more warning below that happens after other possible errors and a return. However, I think it may be okay to move that check up here to make it simpler.

Also, maybe make comment that reading okay up to this point since about to do more tests.

LouisLu705 replied:
I've moved the comment to the final place when all checks have been completed. As for the empty errMsg it should be the case that an errMsg can't be empty with readingOK not false unless the errMsg is a warning. I've added an if statement to handle this warning case here. The logic now is that:

all checks pass
 if (! (errMsg === '')){
     log (errMsg);
} 
I think this should satisfy any possible warnings and in the case there is a real error it will also be logged since all errors are couple with readingOK === false which is checked at the end of given reading number.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				}
+			}
+			if (!readingOK) {
+				// An error occurred, for now log it, let the user know and continue
```
We are going to need some value returned so the user knows when values were dropped. This is going to be trickier since we loop over all readeings. Maybe a counter of number dropped?

LouisLu705 replied:
Alright, I've added a loop to iterate over the reading numbers which were dropped. The idea is that in the file we might have 

HEADER ROW
READING VALUE # 1, START TIME, END TIME (READING # 0)
READING VALUE # 2, START TIME, END TIME (READING # 1)
... (READING # 3)
and so on

and if we drop for example READING #0 because the data is cumulative and this is the first-ever data added to the db we let the user know:

Dropped Reading #0

and also the user will know ahead of time when parsing the data 

"Error Parsing Reading # 0. Reading value gives READING VALUE # 1 with error message: The first reading must be dropped when dealing with cumulative data."

So the full log might be something like,

"Error Parsing Reading # 0. Reading value gives READING VALUE # 1 with error message: The first reading must be dropped when dealing with cumulative data."

Dropped Reading # 0

Now I don't know how this will work with other meter formats but I think it should be fine since ultimately each meter will contain readings and it should just be a matter of how we identify the readings which in my example is in canonical order of the CSV. Also this implicitly counts how many erros occur so we can do something with counting errors if necessary as well.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+						`ROW ${index - readingRepetition}. REJECTED ALL READINGS`);
+					return [];
+				}
+				// To handle cumulative readings that resets at midnight
```
I don't think this check is for midnight but in the allowed time range.

LouisLu705 replied:
You are right. I have updated this comment.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+						meterReading = meterReading2;
+					}
+					else {
+						//cumulativeReset is not expected so throw an error
```
This does not throw an error but logs it and causes the reading to be ignored.

LouisLu705 replied:
Ok thank you, I've updated this so it returns an empty array which is equivalent to processing no readings. I think this makes sense since if there are unexpected net negative meter readings that cannot be handled it would be better to not insert any into the db and let the user know so they can check their data.

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				if (readingOK) {
+					if (Math.abs(prevReading.endTimestamp.diff(prevReading.startTimestamp) - endTimestamp.diff(startTimestamp)) 
+						> Tlen && !isFirst(prevReading.endTimestamp)) {
+						errMsg = 'The previous reading has a different time length than the current reading.';
```
include that this is a warning.

LouisLu705 replied:
Done

huss commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -0,0 +1,169 @@
+				// An error occurred, for now log it, let the user know and continue
+				log.error(errMsg);
+				readingOK = true;
+				if (isCumulative && !onlyEndTime || onlyEndTime && !isCumulative || isCumulative && onlyEndTime) {
```
This is complex logic. Please add a comment to what it is trying to do. Then I can decide if it is correct.

I wonder about the second two or clauses since they both have onlyEndTime and check the other value and not the other value.

LouisLu705 replied:
I've added this comment to make it clear why this logic exists.

/*
If the data is cumulative then regardless of if it comes with end timestamps only or both end timestamps and start timestamps the first reading ever should become the previous reading. This is necessary because there are no previous readings in the db yet so we must drop the first point ever and use this first point as the first previous reading in order to begin calculating net readings since the data is cumulative.

If the data is not cumulative but there are only end timestamps then we still must drop the first reading ever in order to use that reading as the first start timestamp for the next reading. All following readings can then use the previous end timestamps as the current readings start timestamp until all further readings have been processed. 
*/

I think now though after writing this comment out, it may make more sense to instead use isFirst() to check the first reading since this is the only case this logic is necessary.

huss reviewed (changes requested):
I thank @LouisLu705 for good work on a feature with changing needs. I know this is only in a draft state but I wanted to get in some comments now to keep the process moving. @LouisLu705 and I have talked about moving the files in src/server/test/web/csvPipeline/ to be similar to the test/db/data/ directory and excluding them from the header check to avoid that issue. Some CI test issues are due to including the upcoming CSV upload work and will be dealt with in that upcoming PR.  I note that I have only done very limited testing of the pipeline and need to do more types of CSV. Finally, this PR will not be merged until the CSV upload PR has cleared due to dependency on that code.

LouisLu705 committed:
Merged newest version of csvPipeline from Davin.

LouisLu705 committed:
Added comments and removed generatedTestFiles.

LouisLu705 committed:
Merged newest version of developement.

LouisLu705 committed:
Changed first previous reading to take the last value from the meter.

LouisLu705 committed:
Changed pipeline to use Meter model to take in first previous reading data.

LouisLu705 committed:
Added comments.

LouisLu705 committed:
Expanded js documentation for processData(). Also added comments to clarify complex logic.

LouisLu705 committed:
Merged Davin's latest csvPipeline branch.

LouisLu705 committed:
Added TODO for redux state.

LouisLu705 committed:
Added more output to let users know what readings are dropped.

LouisLu705 committed:
Fixed bug with not resetting errMsg.

LouisLu705 committed:
Cleaned up comments. Also fixed duplicate fs.unlink issue where we delete the uploaded csv file multiple times.

LouisLu705 committed:
Fixed npm quotemark issue.

LouisLu705 committed:
Removed trailing whitespace.

LouisLu705 committed:
Fixed query parsing issue. Fixed duplicate handling issue. Added a parameter to take in timeSort. Added more comments.

LouisLu705 committed:
Added indent tabs to processData to satisfy node tests.

LouisLu705 commented:
Thanks for all the comments. I have resolved most of them and left a few open since I'm not sure if some things are satisfactory as is. Some new things to note:

The pipeline now uses the previous values from the meter table to form the first previous reading which slightly increases the time it takes to parse a CSV file. The difference in time was negligible on my local machine but more testing would be nice. I did not do a heavy stress test with at most having ~20 meters in the meter table.

I also added a new parameter to take in timeSort. I noticed for the metasys tests that the CSV files were sorted in descending order instead of the expected ascending order which threw off cumulative values. In order to fix this, I've taken in the timeSort parameter to determine what to do for cumulative depending on ascending vs descending timeSort.

Also, some other things I noticed were incorrect parsing of query parameters. For example, we expected readingRepetitions (which is the number of duplicates) to be a number value but it was instead a string. Similarly, the CSV API was using the refreshReadings parameter check as a boolean when it was instead a string. I've changed it so that the tests pass and the functionality works as intended based on existing comments but there may be other instances where the query parameters are not behaving as expected.

I've also slightly changed how the CSV API deletes files. From testing it seemed to be that the files were being deleted multiple times because there is an fs.unlink call in the CSV API which is called twice regardless if the file was gzipped or not (due to the structure of the code), and also once more in the UploadReadings.js. This was causing the file to be deleted once after it was created, once more because it thought the file was unzipped, and then once more in upload readings after the CSV file is inserted into the db. It seems that even if the file is deleted multiple times it is fine in terms of the data being correctly inserted into the db but I have made some slight changes that hopefully fixes this issue.

Also, I've changed how the pipeline handles duplicates so instead of the complex logic previously it is instead a single for loop which increments a counter by the duplicate value starting from 0. This works for all my tests but more testing is desirable. I noticed in the metasys test that with a duplicate value of 2, the first reading is actually only seen once and then all the other readings are duplicated and seen twice. I believe that the for loop will still work in such a case but it has not been tested on very large duplicate values. So for example,

Unique first reading, start time, end time (0)
second reading, start time, end time (1)
duplicate second reading ... (2)
third reading, start time, end time (3)
duplicate third reading ... (4)
...
The idea is we start at 0 and if readingRepetition = 2 then we have itr 0 at 0. itr 1 at 0+2=2. itr 2 = 2+2 = 4.... and so on. If the readingReptition is 1 (there are no duplicates) then it will increment by 1 which also works. I haven't been able to come up with an example where this doesn't work yet.

Finally, I added a bunch of comment updates as well as a slight change to the output errors at the end. Big thanks to @huss for the detailed comments and observations.

If anything is confusing or unclear let me know and I will explain in further detail.

huss committed:
formatting changing

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/LouisLu705/629

huss committed:
row error msgs start at 1 not 0

huss committed:
add row & values to reset issue msg

huss committed:
modify developer test data script

The script was updated to use the new CSV input and pipeline.
It was also cleaned up and info message added.
Also create need directory and place file so will be in repo.

huss committed:
add line # to warning message

huss committed:
listed readings dropped start at 1 not 0

huss committed:
use \n for new lines in echo

huss committed:
upgrade cumulative reset

This makes a series of changes for cumulative reset:
- Used reset values from route rather than default/set values.
A TODO noted this but now it is done.
- If no reset value given then used DB ones.
- Updated functions, calls & tests throughout code for new reset values.
- Use moment.duration for the allowed reset times. This seems better
since there is no date associated with them.
- Changed calculation of reset moments but not idea.
- Use default moment format since only the values from the DB are used
this way. Other way seemed to cause issues at times.
- Added so reset can cross a day per newer specifications.
Also did:
- minor formatting

huss committed:
accept/use reading length variation

- Modify code so this input is used rather than default value.
- Update calls including testing for new parameter.
- Use new variable name rather than Tlen used before.
- Fix issue that not using meter values in call to loadArrayInput
for obvius.

huss committed:
add usage of gap check between readings

- The type for gap in DB was made real and named reading_gap.
- Note this replaced reading_length. The existing
reading_variation checks this and now that we store the last reading
on the meter it can check all but the first value.
- Meter creation has teh gap and default values added to constructor.
- All usage and calls where updated for the new gap.
- The order of function arguments was updated to be more consistent
with other code. A few issues with usage were also corrected.
- Fixed message on gap when not first reading.
- Added gap to message.
- Fixed up Metasys items even though hopefully going away.

huss committed:
missed file from last commit

huss committed:
fixes to CSV upload

- The file name for readings was accidently named meters so corrected.
- The message on meter & reading CSV upload were made parallel
and nicer.
- Bugs with naming of reading length and gap were fixed.
- Message when cumulative and negative readings were made clear.
First, row begins at 1.
Second, if first row then value came from meter.

huss committed:
information on pipeline testing

The files & script were added and README was updated to describe the
pipeline tests run to check the PR. These are not everything desired but
a good start and plan to convert to automated testing.

An old, unneeded README was removed.

huss committed:
Use lengthGap not length & bug for web values not set

- In the original CSV upload, there was a length parameter but now it is
the gap so named changed to reflect this.
- The check to set a value not provided worked for curl but the web
page send and empty string so checks updated.
- Update pipeline tests to check if use parameters when DB values set.

huss committed:
add decreasing option for time sort

huss committed:
add cumulative decending test

huss committed:
log if meter created & if already exists

huss committed:
add more logging info

When a warning or error is generated, the status is now dumped.
As noted, the order can be off with multiple message.

huss committed:
fix formatting issue

huss committed:
add items in CSV upload

- Allows for duplication, end only & time sort for upload.
- Modified DB, meter & testing for these.
- Made so parameters not provided default to DB or default if no
DB value.
- Added meter option on time sort that is to use meter value.
- Made time sort from meter be default on web page.
- Note that end only option not yet on web page.
- cumulative & reset not yet able to use meter value.

huss committed:
handle cumulative & reset DB values

- This changes these value so it is a dropdown menu with three choices
for true, false and meter or default value (default choice)
- Puts back test for duplication being an integer but adds can be empty
string as needed with new system from previous commit.

huss committed:
add end only to web page + other

- Redid pipeline logic for decreasing data by reversing loop rather than
checking inside pipeline.
- Reordered values so the web page, function calls, etc. are
in same order.
- Updated log with readings so show correct values.
- Changed Obvius to use meter values rather than generic defaults.
This will not make a real change unless the meter values are edited.
- Minor fix to test code including Metasys test that assumed ordering when only
looking at one value from DB.
- Fixed a couple of JSDoc locations where wrong params shown.
- Tried to make Select/Set name usage consistent for Redux functions.

huss committed:
send back log output when do curl

- Messages went to the console when doing CSV upload for readings but
were not returned as part of requst. Now they come back as HTML.
This will make them look good if in a web browser or each to format
by putting into one.
- Failure is not correctly shown for any dropped reading whereas
before it was only if an exception occurred.
- It is possible that some warning/message were lost before when
another error occurred but will certainly show now.
- Values from the user are not html escaped but it is assumed done
before getting to the process step. This was now noticed but left.
- Note this does not fix what is returned for web page CSV upload nor
does it address messages for meter CSV upload.
- It also does not address other calls to the modifed methods.
This likely why one Metasys test is failing. Fix soon.

huss committed:
remove Metasys code

The new pipeline & CSV input offers all the input options of the old
Metasys code. Maintaining this older code no longer made sense so
it has been remove along with the associated tests. Note the old version
and the newer one when the pipeline was first begun are removed.
The one item lost is the fact that the code edited the incoming data to
remove the units on the reading. This can easily be done by a site or
another layer of code that feeds into the new pipeline.
Note the Metasys type meter is left for historical reasons.

huss committed:
add file of expected pipeline test output

This is to make doing a diff easier and to know what to expect.
This is viewed as temporary until the automated tests are added.

huss committed:
remove CSV stream & stream; fix Obvius issue

Stream input for CSV is now well out of date given the new pipeline.
Time tests show that with regular CSV input loading it all into memory
works fairly quickly. If you have a non-server with less memory where
the amount available for JS is less than about 2GB then a file too much
over 6MB/100k points may run out of memory. This can be fixed by
increasing the memory available to JS/Node. Another fix is to split
the file into sections of smaller size. This is not expected to
happen on a live server with reasonable memory/setting. Given all of
this, it was decided to remove the stream option rather that upgrade it.

While checking for stream in code, found a number from the old meter and
reading drop zone that was replaced with the new one. All that code for
web pages, routes, translations, etc. were deleted. Also found two
stream package that are not used (one seems old issue).

Note the changes pass all but one test but have yet to be
run on live MAMAC
meters (that used to use stream but not really necessary since come as
1440 points at a time). This will be done before the PR is updated.
The failing test is believed to be caused by how errors are handled
in the new pipeline. Getting all the calling functions to correctly
handle new return values should fix it and will be done soon.

package-lock.json needs updating for package changes. Will do this
later when online.

The Obvius example upload was failing because identifier was made unique
in DB. Changed so identifier appends unique ID.

huss committed:
fix but on unenabled MAMAC meters

The update script was trying to get data from meters that were
not enabled.

huss committed:
undo last change

I was working too fast. There is not a bug here.

huss committed:
get mamac to use new pipeline

- Also removed old mamac pipeline file and test
- mamac readings generally a whole number but removed round since
OED now deals with real values.
- The loading of all mamac meters at one time on update was removed
since the pipeline does one meter at a time.
- Pipeline no longer specifies how moment should parse date/time.
Testing found it worked fine without except day first but that was not
handled anyway.
- Removed updateMetersTests since we are not doing the rejection as done
before. I hope this is  not overlooking an issue where one update fails
and another one does not work correctly.

huss committed:
begin Obvius with new pipeline + fixes

- The data is sent through the pipeline rather than direct adding to DB.
The readings are still done one at a time. More work needed but
basic functioning.
- Made await array load await as it should.
- Change so always add last reading to meter instead of if success.
Also, await so that fast next request for same meter works.

huss committed:
limit size of messages returned

If the pipeline is used with a large file with lots of errors, the
returned message could be quite large. This limits it to about 75k
and 75 messages. The log still holds all the messages.

huss committed:
add meter name to message + other

- If the pipeline is called many times quickly, it can be hard to tell
which meter is involved. Thus, the meter name was added to
all the messages.
- Obvius log file tests failing because 1970-01-01 as the date
and this is the same as moment(0) that is used on meter when
there is no data. Also, removed overlapping test since this was
when assumed each reading was 1 hour. Finally, removed blank
line for synthetic CSV data because lead to null reading.
- 1970 is used in a other places but none seem to involve adding
readings to DB so probably okay.
- Fixed the issue with multiple return values from appendMsgTotal
so don't need different named variables.
- Fixed some check/lint issues.
- Updated expected output for new one.

huss committed:
single Obvius log upload + others

- There were two functions to upload Obvius data. insertObviusData
was not used anywhere so deleted. Anything needed was transferred
to loadLogfileToReadings and is only one now.
- Obvius upload now does each meter as a batch rather than each
reading.
- Obvius readings uploads now honor meter values for upload
parameters. Changed meter creation for config and log so
end only is true.
- Changed all meter creations that were automatic to set the note to
how done and a date/time stamp.
- Fix bug where default meter value was ascending but pipeline has
increasing

huss committed:
simple fix

huss committed:
failure if cumulatve/cumulative reset inconsistent

huss committed:
pipeline msgs returned to web page

- Curl requests were getting the message and now ones from the web page
are getting the same output. It is displayed as an alert so it stays on
the screen and is easy to copy.
- Moved curl message return until after done with deleting files and
refresh. This should make it sooner to when ops actually done.
- I had some issue using <string> instead of <void> so left it as void.
Maybe it can be figured out at a later point.

huss committed:
clarify where ignoring pipeline return values

Some automated calls ignore the values. Also in tests for now.

huss committed:
Get update to work for readings

- Also fixed type of readingsCumulative

huss committed:
meter update & web meter msgs

- Updates for meters now works with testing added for this.
- The same alert style messages are shown for web page requests as were
done for meter curl requests.
- Note there is a meterName for curl meter requests but not on the meter
web page. This may be fixed later by either adding edit to meter page
and/or adding meter name to csv meter page.

huss committed:
add other meter type

We OED's generalization of adding meters/readings, we are beyond the
meter types known. Thus, an "other" type has been added. It is used when
a meter is automatically created but the type is not know. A site can
also use this type.
A few tests were changed to use this type to be sure it works.

huss committed:
make DB start/end timestamp be moment(0)

Originally we planned to user Jan 1, 0001 as the default start/end time
on a meter. The code uses moment(0) which is 1970-01-01 00:00:00.
This makes the DB consistent with the code even though it is unlikely
to ever see the DB default because the code sets a default value on
meter creation.

huss committed:
clean up todos in pipeline

- Almost all were either dealt with in this commit or were already okay.
- Moment conversions were set to strict so unlikely a bad value
will get by.
- Invalid dates or readings (not a number) are now caught with
appropriate message where all reading are rejected.
- Tests were added for all the new checks.

huss committed:
modify pipeline comments

I carefully read the new pipeline code. Mostly just added or edited
comments to make it clearer.
Did one small change that does not change any results but makes
code safer.

huss committed:
adds meter name input + fixes

- The CSV meter input page now accepts a meter name. This is used rather
than value from CSV in same way a curl from previous code.
- Changed a few strings to use MODE instead
- Fixed expected output for actual reading value.
- Removed stray character in pipe41 CSV.
- Fixed pipeline README so descriptions correct.

huss commented:
First, I want to thank @LouisLu705 for all the work on the pipeline. I don't have any additional comments on their code.

There are some CodeQL issues that will need to be addressed at some point. It might be after this PR is merged.

I'm pushed a lot of changes/commits to the PR. I kept making changes to finish tasks and it got somewhat out of hand where changes in related code outside the original PR is now included. The positive is that the pipeline and related code should now be fully integrated and working with all options. Given the scope of the work, my changes should be reviewed before they are merged with the PR. The changes are discussed in each commit but here are the highlights:

- The curl & web page now have the same abilities. This required many changes including unspecified values now default to what is stored on the meter (when appropriate) so Fixes #641; the database was modified so it has values for every value that makes sense in the pipeline; some new abilities to handle all the options; update option now correctly works with DB so functional.
- Messages from the pipeline are now returned and displayed on a web page popup or as the return HTML for curl requests. The messages were enhanced to make them more useable. A maximum size for the returned message so won't report too many issues but all go to the log.
- MAMAC and Obvius now go through the pipeline. The duplicate capability from previous work is now gone. Previous code for the MAMAC drop zone was removed. Note that the reading length assumption of the old Obvius code is now gone as pipeline can figure out reading length. Fixes #510. Partly addresses #564.
- A new meter type of other was added so automatically created meters of unknown origin have a reasonable type. Fixes #565.
- The stream input option has been removed. The new pipeline seems to handle reasonable size CSV files so stream was not worth upgrading. Note testing shows that large files can fail if you don't give JS enough memory. This seems more a developer rather than production server issue.
- The old Metasys import code was removed. The new pipeline does everything it did and more. Fixes #512. Fixes #515 as making it obsolete.
- The developer script for inputting test data now goes through the pipeline.
- Made DB have same default start/end date/time as moment(0) so they are consistent. Also changed some of the tests that happened to use this date/time and was causing failure. Many were left as not causing issue.
- A script with test CSV files is included that runs many tests against the pipeline. Note the hope is that many can be converted to Mocha/Chai tests so they become part of the continuous integration.
- Many other changes including removal of obsolete tests, few bug fixes, better error checking on some parameters, enhanced commenting, etc.

huss committed:
expand csvPipeling testing

- Add some tests for cumulative
- Add test for start before end
- minor fixes to processData.js

huss committed:
avoid removal of docker container

The --rm on the docker command caused running versions of OED to stop.
As a compromise, this was removed but it means the container
stays around.

lindavin reviewed (approved):
This looks really good. I'm very glad that mamac and obvius utilize the same pipeline code now. I don't see any obvious JS errors. I also like the addition of the enums TimeSortTypesJS and BooleanTypesJS in validateCSVUploadParams. In some places we checked if a variable is equal to string 'true', for example, line 150 of /src/server/routes/csv.js. I think this could be changed to utilize BooleanTypeJS as a later TODO.

huss committed:
use enum per PR review

Changes code to use TimeSortTypesJS & BooleanTypesJS instead of
hardcoded values used before.

huss commented:
After my review of the work by @LouisLu705 and the review of my changes by @lindavin, I think this is ready to merge. The CodeQL issues will be addressed at a later time.
